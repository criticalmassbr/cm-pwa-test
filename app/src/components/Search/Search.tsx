import * as React from 'react'
import styled from 'styled-components'
import { useLazyQuery, gql } from '@apollo/client'
import { Redirect } from 'react-router-dom'
import { Link } from 'react-router-dom'

const Heading = styled.strong`
  font-size: 2rem;
  margin: 0.75em 0 1em;
`

const Wrapper = styled.div`
  margin-bottom: 0.5rem;
  align-content: center;
  justify-content: center;
`

const Input = styled.input`
  margin-left: 1rem;
  type: text;
  id: "search";
  placeholder: "search";
  size: 10rem;
  value: "";
  height: 2rem;
  border-radius: 16px;
`

const Button = styled.input`
  value: "Buscar";
`
interface Friend {
  _id: string,
  name: string,
  email: string,
  picture: string,
  company: string,
  age: number,
  eyeColor: string,
  index: number,
  friends: [Friend]
}

interface QueryData {
  list: Friend[]
}

const query = gql`
  query ($name: String){
    list (name: $name) {
      _id
      name
      email
      company
      age
      eyeColor
      picture
      index
      friends {
        name
        email
        company
        age
        eyeColor
        picture
        _id
      }
    }
  }
`

const Search = () => {
  const [runQuery, { data, loading }] = useLazyQuery<QueryData>(query) 
  if(loading) {
    return <div>Carregando...</div>
  } 
  if(data && data.list[0]) {
    const friend = data?.list[0]
    return (
      <Redirect to={{ pathname: `/${friend._id}`, state: { friend }}}></Redirect>
    )
  }
  if(data && data.list.length == 0) {
    const friend = data?.list[0]
    return (
      <Redirect to="/"></Redirect>
    )
  }
  
  const onSearch = (event: React.MouseEvent<HTMLInputElement, MouseEvent>) => {
    event.preventDefault()
    const input = document.getElementById("search") as HTMLInputElement
    const name = input.value
    runQuery({ variables: { name }})
  }
  
  return (
    <Wrapper>
      <Link to="/">
        <Heading>MySocial</Heading>
      </Link>
        <Input id="search" />
        <Button type="submit" onClick={onSearch} />
    </Wrapper>
  )
}

export default Search