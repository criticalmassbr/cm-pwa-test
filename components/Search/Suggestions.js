import { SuggestionsSearch } from "./SearchStyle";

const Suggestions = (props) => {
  return (
    <SuggestionsSearch>
      {props.results.map(({ name }, index) => (
        <li key={index + 1} onClick={e => props.searchArtists(props.textSearch, e)}>
          <p className="mt-6px bg-none tx-darkgrey fn-s14px ln-h19px tx-dn">
            {name}
          </p>
        </li>
      ))}
    </SuggestionsSearch>
  );
};

export default Suggestions;
