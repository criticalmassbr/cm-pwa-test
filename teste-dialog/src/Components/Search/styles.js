import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  padding: 50px;
  width: 100%;
`;

export const DivSearch = styled.div`
  max-width: 100%;
  min-width: 290px;
  flex-direction: column;
  display: flex;
`;
export const LabelSearch = styled.div``;
export const InputSearch = styled.input`
  background: #fff;
  border-radius: 20px;
  height: 35px;
  border: #fff;
  width: 100%;
  padding-left: 40px;
  padding-right: 15px;
  font-family: sans-serif;
`;

export const AlignIcon = styled.div`
  float: right;
  right: 0;
  margin-top: -30px;
`;

export const FormContainer = styled.div`
  width: 100%;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
