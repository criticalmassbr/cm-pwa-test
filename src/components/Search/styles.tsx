import { Search } from '@styled-icons/boxicons-regular/Search';
import styled from 'styled-components';

export const Center = styled.div`
  display: flex;
  justify-content: center;
`;

export const Container = styled.div`
  font-size: 1.5em;
  font-weight: bold;
  line-height: 1em;
  position: relative;
`;

export const SearchIcon = styled(Search).attrs(attr => ({
  size: attr.size || '1.5em',
}))`
  position: absolute;
  top: 0.7em;
  left: 0.9em;
  width: 1.25em;
`;

export const Input = styled.input`
  border-radius: 0.75em;
  box-shadow: 0 0 3px #ccc, 0 10px 15px #ebebeb inset;
  margin: 1em;
  padding: 0.25em 1em;
  text-indent: 1em;
  width: 50vw;
  font-size: 0.75em;
`;
