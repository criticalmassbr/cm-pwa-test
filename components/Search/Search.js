import { useState, useEffect, useCallback } from "react";

import * as GridStyle from '~/styles/Grid';
import {
  FormSearch,
  InputSearch,
  FlagSearch
} from "./SearchStyle";

import Suggestions from '~/components/Search/Suggestions';

import axios from "axios";

const searchArtists = (props) => {
  const [textSearch, setText] = useState("");
  // eslint-disable-next-line no-unused-vars
  const [isVisible, setVisible] = useState(false);
  const [inFocus, setFocus] = useState(false);

  const [state, setState] = useState({
    query: '',
    results: []
  });

  const handleSearch = useCallback(async (text) => {
    if(text.length > 2) {
      const query = encodeURIComponent(`{
        queryArtists(byName: "${text}") {
          name
          id
          image
          albums {
            name
            id
            image 
          }
        }
      }`);
  
      try {
        axios(`https://spotify-graphql-server.herokuapp.com/graphql?query=${query}`)
          .then(({ data }) => {
            setState({
              results: data.data.queryArtists
            })
          })
      } catch (error) {
        console.log(error);
      }
    }
  });

  const onClickSearch = () => {
    redirectSearch(textSearch);
  };

  const redirectSearch = (text) => {
    if (!text || text.length < 3) return;
    handleSearch(text);
  };

  function changePropsVisibleSearch(param) {
    props.changeVisibleSearch(param);
  }

  useEffect(() => {
    let searchTimer = setTimeout(() => {
      handleSearch(textSearch);
    }, 150);
    return () => clearTimeout(searchTimer);
  }, [textSearch]);

  return (
    <>
      <FormSearch className="w-100 flex">
        <GridStyle.Row className="no-gutters w-100">
          <GridStyle.Col className="no-margin no-padding" general={12}>
            <GridStyle.Row>
              <GridStyle.Col className="no-margin" general={9}>
                <InputSearch
                  placeholder="Search your Artists"
                  value={textSearch}
                  name="query"
                  className="fn-s14px ln-h19px"
                  onKeyPress={(e) => e.key === "Enter" && onClickSearch()}
                  onChange={(e) => setText(e.target.value)}
                  onBlur={() => !inFocus && changePropsVisibleSearch(false)}
                  onFocus={() => changePropsVisibleSearch(true)}
                />
              </GridStyle.Col>

              <GridStyle.Col className="no-margin" general={3}>
                <FlagSearch />
              </GridStyle.Col>
            </GridStyle.Row>
          </GridStyle.Col>

          <div
            className="black-block"
            onClick={() => setVisible(false) && changePropsVisibleSearch(false)}
          />

          {typeof state.query === "undefined" &&
            state.results &&
            state.results.length >= 1 &&
            props.isVisibleSearch === true && (
              <GridStyle.Col
                className="no-margin no-padding"
                general={12}
                onMouseEnter={() => setFocus(true)}
                onMouseLeave={() => setFocus(false)}
              >
                <Suggestions
                  results={state.results}
                  searchArtists={props.searchArtists}
                  textSearch={textSearch}
                />
              </GridStyle.Col>
            )}
        </GridStyle.Row>
      </FormSearch>
    </>
  )
};

export default searchArtists;
