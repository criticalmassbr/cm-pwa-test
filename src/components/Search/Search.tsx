import React, { useEffect, useRef, useState } from 'react';
import { Center, Container, Input, SearchIcon } from './styles';

type Props = {
  handleSearch: (term: string) => void;
};
function Search({ handleSearch }: Props) {
  const [term, setTerm] = useState('');

  const debounce = useRef(0);
  const _onChange = ({ target: { value } }: any) => {
    if (debounce.current) clearTimeout(debounce.current);
    debounce.current = setTimeout(() => {
      setTerm(value);
    }, 500) as any;
  };

  const inputRef = useRef<HTMLInputElement>(null);
  const _onMouseEnter = () => {
    if (inputRef.current) inputRef.current.focus();
  };

  useEffect(() => {
    if (!handleSearch || !term) return;
    handleSearch(term);
  }, [term, handleSearch]);

  return (
    <Center>
      <Container>
        <SearchIcon />
        <Input
          ref={inputRef}
          name="search"
          type="text"
          placeholder="Search"
          onChange={_onChange}
          onMouseEnter={_onMouseEnter}
        />
      </Container>
    </Center>
  );
}

export default Search;
