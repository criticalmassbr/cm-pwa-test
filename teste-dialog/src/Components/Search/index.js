import React from "react";
import { MdSearch } from "react-icons/md";

import { Container, DivSearch, InputSearch, AlignIcon } from "./styles";

export default function Search() {
  return (
    <Container>
      <DivSearch>
        <InputSearch placeholder="Busque mÃºsicas, artistas, podcasts..." />
      </DivSearch>
      <AlignIcon>
        <MdSearch size={26} color="#000" />
      </AlignIcon>
    </Container>
  );
}
