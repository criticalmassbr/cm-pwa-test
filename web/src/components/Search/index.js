import React, { useState } from 'react';
import { SearchInput } from './styles';

export default function Search( {  setUserSearch } ) {
  const [value, setValue] = useState('');

  return (
    <SearchInput>
      <input
        value={value}
        onChange={e => setValue(e.currentTarget.value)}
        type="text"
        placeholder="search"
        onKeyDown={event => {
          if (event.key === 'Enter') {
            setUserSearch(value)
          }
        }}
      />
    </SearchInput>
  )
}


