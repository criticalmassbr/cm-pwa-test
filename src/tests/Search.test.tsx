import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../components/Search/Search';

test('renders artist cover and name', () => {
  const handleSearchMock = jest.fn(() => {});
  const component = <Search handleSearch={handleSearchMock} />;
  const searchComponent = renderer.create(component);
  const tree = searchComponent.toJSON();
  expect(tree).toMatchSnapshot();
});
