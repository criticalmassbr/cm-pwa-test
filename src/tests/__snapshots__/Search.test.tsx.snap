// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders artist cover and name 1`] = `
<div
  className="sc-AxirZ VxIXc"
>
  <div
    className="sc-AxiKw eqNJIm"
  >
    <svg
      aria-hidden="true"
      className="sc-AxjAm gXIMwt sc-AxhCb fQTWBy"
      fill="currentColor"
      focusable="false"
      height="1.5em"
      viewBox="0 0 24 24"
      width="1.5em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 18a7.952 7.952 0 004.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0018 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
      />
    </svg>
    <input
      className="sc-AxhUy ikFBxM"
      name="search"
      onChange={[Function]}
      onMouseEnter={[Function]}
      placeholder="Search"
      type="text"
    />
  </div>
</div>
`;
